src/k2/dto/LogStream.h

```diff
@@ -0,0 +1,111 @@
+/*
+MIT License
+
+Copyright(c) 2020 Futurewei Cloud
+
+    Permission is hereby granted,
+    free of charge, to any person obtaining a copy of this software and associated documentation files(the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and / or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions :
+
+    The above copyright notice and this permission notice shall be included in all copies
+    or
+    substantial portions of the Software.
+
+    THE SOFTWARE IS PROVIDED "AS IS",
+    WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+    DAMAGES OR OTHER
+    LIABILITY,
+    WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+    SOFTWARE.
+*/
+#pragma once
+
+namespace k2 {
+namespace dto {
+
+struct MetadataRecord{
+    String plogId;
+    uint32_t sealed_offset;
+    K2_PAYLOAD_FIELDS(plogId, sealed_offset);
+    K2_DEF_FMT(MetadataRecord, plogId, sealed_offset);
+};
+
+// Request to create a Metadata Log Stream Record 
+struct MetadataPersistRequest {
+    String partitionName;
+    uint32_t sealed_offset;
+    String new_plogId;
+    K2_PAYLOAD_FIELDS(partitionName, sealed_offset, new_plogId);
+    K2_DEF_FMT(MetadataPersistRequest, partitionName, sealed_offset, new_plogId);
+};
+
+struct MetadataPersistResponse {
+    K2_PAYLOAD_EMPTY;
+    K2_DEF_FMT(MetadataPersistResponse);
+};
+
+struct MetadataGetRequest {
+    String partitionName;
+    K2_PAYLOAD_FIELDS(partitionName);
+    K2_DEF_FMT(MetadataGetRequest, partitionName);
+};
+
+struct MetadataGetResponse {
+    std::vector<MetadataRecord> records;
+    K2_PAYLOAD_FIELDS(records);
+    K2_DEF_FMT(MetadataGetResponse, records);
+};
+
+
+struct LogStreamBaseExistError : public std::exception {
```

Comment:
Exceptions don't make sense to be here. We do not return exceptions in the dto interface - we return status codes. You should move these to the logstream implementation instead, and then in your code convert exceptions to the correct status codes.

It may be even better to write your code so that it operates in terms of Status instead of throwing/catching exceptions

Reply:
changed all throw exception to return status